lib/create-plugin.js

```diff
@@ -3,11 +3,126 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
- 'use strict'
+'use strict'
 
-function createPlugin() {
+function createPlugin(instrumentationApi) {
   return {
+    requestDidStart() {
+      const longestResolvedPath = {
+        depth: 0,
+        formatted: null
+      }
 
+      return {
+        executionDidStart: () => ({
+          willResolveField({info}) {
+            const pathArray = flattenToArray(info.path)
+            const formattedPath = pathArray.reverse().join('.')
+
+            if (pathArray.length > longestResolvedPath.depth) {
+              longestResolvedPath.depth = pathArray.length
+              longestResolvedPath.formatted = formattedPath
+            }
+          }
+        }),
+        willSendResponse(responseContext) {
+          // default. only used if failed to parse
+          let transactionName = '*'
+
+          if (responseContext.operation) {
+            // Resolved operation: parsed and validated, will execute resolvers
+
+            const operationType = responseContext.operation.operation
+            const operationName = responseContext.operationName || '<anonymous>'
+            const longestPath = longestResolvedPath.formatted
+
+            transactionName = `${operationType} ${operationName} ${longestPath}`
+          } else if (responseContext.document) {
+            // Failed validation, fallback to document AST
+
+            transactionName = getNameFromDocument(responseContext.document)
+          }
+
+          setTransactionName(instrumentationApi.agent, transactionName)
+        }
+      }
+    }
+  }
+}
+
+function getNameFromDocument(document) {
+  // TODO: when would there be multiple document definitions?
+  const definition = document.definitions[0]
+  const operationType = definition.operation
+  const operationName = definition.name || '<anonymous>'
+
+  const longestSelectionPath = getLongestSelectionPath(definition)
+
+  const longestPath = longestSelectionPath.join('.')
+  return `${operationType} ${operationName} ${longestPath}`
+}
+
+function flattenToArray(fieldPath) {
```

Comment:
minor: but there are helper functions for flattening path and other things off the 'graphql' module which might be available for the plugin.
https://github.com/apollographql/apollo-server/blob/60d974240769aac3e1c75ff968657b2d3e60e9b6/packages/apollo-tracing/src/index.ts#L3

Reply:
ah cool. although, we might want to avoid relying on being able to import graphql / having a dependency there for something that could change. at least for something this simple. that being said, there might be other things it has we want to use which would change the consideration.